src/mpid/ch4/shm/xpmem/xpmem_progress.h

```diff
@@ -0,0 +1,205 @@
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+#ifndef XPMEM_PROGRESS_H_INCLUDED
+#define XPMEM_PROGRESS_H_INCLUDED
+
+#include <xpmem.h>
+#include "ch4_impl.h"
+#include "../posix/posix_progress.h"
+
+#undef FCNAME
+#define FCNAME MPL_QUOTE(MPIDI_XPMEM_progress_recv)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_progress_recv(int blocking)
+{
+    MPIDI_POSIX_eager_recv_transaction_t transaction;
+    int mpi_errno = MPI_SUCCESS;
+    int i;
+    int ack;
+    int result = MPIDI_POSIX_OK;
+    MPIR_Request *rreq = NULL;
+    MPIDIG_am_target_cmpl_cb target_cmpl_cb = NULL;
+
+    void *data_buf = NULL;
+    void *exp_buf = NULL;
+    xpmem_apid_t apid;
+    MPIDI_XPMEM_head *xpmem_header = NULL;
+    void *p_data;
+    size_t p_data_sz = 0;
+
+    int is_contig;
+    int outcount;
+    int type_size;
+    MPI_Aint position = 0;
+    MPIR_Datatype *dt_ptr;
+    void *am_hdr = NULL;
+    MPIDI_POSIX_am_header_t *msg_hdr = NULL;
+    MPIDI_POSIX_am_header_t ack_msg_hdr;
+    MPIDI_POSIX_am_header_t *ack_msg_hdr_ptr;
+    MPIDI_CH4U_hdr_t ack_am_hdr;
+
+    MPIR_Comm *comm;
+    uint8_t *payload;
+    uint64_t context_id;
+    size_t payload_left;
+    struct iovec iov_left[2];
+    size_t iov_num;
+    struct iovec *iov_left_ptr;
+
+    struct iovec *iov_ptr;
+    long long recv_data_sz;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_PROGRESS_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_PROGRESS_RECV);
+
+    /* Check to see if any new messages are ready for processing from the eager submodule. */
+    result = MPIDI_POSIX_eager_recv_begin(&transaction);
+
+    if (MPIDI_POSIX_OK != result) {
+        goto fn_exit;
+    }
+
+    /* Process the eager message */
+    msg_hdr = transaction.msg_hdr;
+    payload = transaction.payload;
+
```

Comment:
Looks like you copied posix progress code here ? We should not do it. Why do you need it ?